### DISCUSSION TITLE
VSCode NX Console + NixOS

### DISCUSSION THREAD

**Discussion by dsod** [nrwl/nx-console#2051] [General]
After having some issues with the NX Console plugin in VSCode and solving it, I thought that I should share my findings here for anyone else experiencing the same issue.

The issue I had was that the NX Console client failed to start when starting VSCode, and the console gave me the following output:
```
Error: No prebuild or local build of @parcel/watcher found. Tried @parcel/watcher-linux-x64-glibc. Please ensure it is installed (don't use --no-optional when installing with npm). Otherwise it is possible we don't support your platform yet. If this is the case, please report an issue to https://github.com/parcel-bundler/watcher.
```

I verified that the package it tried to import in fact was installed in `~/.vscode-insiders/extensions/nrwl.angular-console-18.17.0/node_modules/@parcel/watcher-linux-x64-glibc`, so I thought that it could be an issue with running the binary. Since the parcel watcher is a C++ binary that uses dynamic linking to load system libraries during runtime, this was likely the case. I did some cleanup:

- Re-installed the NX Console plugin (made sure the folder in `~/.vscode-insiders/extensions/nrwl.angular-console-18.17.0` was properly removed)
- Made sure that the environment variable `LD_LIBRARY_PATH` was _not_ set when starting vscode though the terminal (`unset LD_LIBRARY_PATH`. In my case, a pipewire setting in NixOS had set the variable, which could cause issues in steps further down)

Now I started getting another error: `Error: libstdc++.so.6: cannot open shared object file: No such file or directory`

Verified the issue by running `ldd ~/.vscode-insiders/extensions/nrwl.angular-console-18.17.0/node_modules/@parcel/watcher-linux-x64-glibc/watcher.node` which gave me:
```
linux-vdso.so.1 (0x00007ffedd7e7000)
**libstdc++.so.6 => not found**
libm.so.6 => /nix/store/ksk3rnb0ljx8gngzk19jlmbjyvac4hw6-glibc-2.38-44/lib/libm.so.6 (0x00007b62ef13e000)
libgcc_s.so.1 => /nix/store/d8w5qfswmgxcjqwnmqw2v9r8amrvdlpb-xgcc-13.2.0-libgcc/lib/libgcc_s.so.1 (0x00007b62ef119000)
libpthread.so.0 => /nix/store/ksk3rnb0ljx8gngzk19jlmbjyvac4hw6-glibc-2.38-44/lib/libpthread.so.0 (0x00007b62ef114000)
libc.so.6 => /nix/store/ksk3rnb0ljx8gngzk19jlmbjyvac4hw6-glibc-2.38-44/lib/libc.so.6 (0x00007b62eef2b000)
/nix/store/ksk3rnb0ljx8gngzk19jlmbjyvac4hw6-glibc-2.38-44/lib64/ld-linux-x86-64.so.2 (0x00007b62ef222000)
```

Right, so the NixOS link-loader is missing a C++ library. Here are two ways of solving this:
1. Install the plugin through a Nix derivation which includes the c++ library in the binary wrapper (i.e. through the excellent [nix-vscode-extensions](https://github.com/nix-community/nix-vscode-extensions) library). This was not working for me either (back to the Module not found issue)
2. Utilize [nix-ld](https://github.com/Mic92/nix-ld) link-loader which allows me to run dynamic libraries like a regular linux distro

I already had [nix-ld-rs](https://github.com/nix-community/nix-ld-rs) set up on my system. I used the maintainers dotfiles for reference: https://github.com/Mic92/dotfiles/blob/main/nixos/modules/nix-ld.nix. `nix-ld` will however not automatically fix the parcel binary for us, because it is run from node `nix-ld` won't be used. I am still exploring the best way to scope vscode extensions to using the `NIX_LD_LIBRARY_PATH` without setting the it for subprocesses, like shells within VSCode. Right now, I am doing this:

1. Run VSCode with `LD_LIBRARY_PATH=$NIX_LD_LIBRARY_PATH` code-insiders ...`
2. Unset the `LD_LIBRARY_PATH` when a shell open. I use `direnv` to automatically load development environments, so it's as easy as adding `unset LD_LIBRARY_PATH` into the `.envrc` file

The permanent solution should probably be to contribute to https://github.com/nix-community/nix-vscode-extensions and make sure that the libstdc++.so.6 library is bundled with VSCode whenever the NX Console extension is installed.

Feel free to provide feedback or discuss other solutions to the issue below.

**Comment by MaxKless** [comment]
Thanks so much for the in-depth research & write-up. This is surely going to help a lot of people!

**Comment by clemenscodes** [comment]
Thanks a bunch! Using the `nix-ld` approach, I ended up creating a home-manager desktop entry which handles this for me (on wayland).
```nix
{...}: {
  xdg = {
    desktopEntries = {
      codium = {
        name = "VSCodium";
        type = "Application";
        categories = ["Utility" "TextEditor" "Development" "IDE"];
        comment = "Code Editing. Redefined.";
        exec = ''sh -c "LD_LIBRARY_PATH=\\$NIX_LD_LIBRARY_PATH codium --enable-features=UseOzonePlatform --ozone-platform=wayland --enable-features=WaylandWindowDecorations"'';
        genericName = "Text Editor";
        icon = "vscodium";
        mimeType = ["x-scheme-handler/vscode"];
        noDisplay = false;
        startupNotify = true;
        terminal = false;
      };
    };
  };
}
```

**Reply by dsod** [reply]
That is great! Glad it helped.

Keep in mind that any sub-process, i.e. a terminal inside VSCode will have the `LD_LIBRARY_PATH` variable set in this setup. It can mess up other CLI tools / binaries installed in your NixOS installation. That is why I am unsetting it again for terminals spawned inside vscode.

**Reply by clemenscodes** [reply]
Yeah, for that I'll also use `direnv`, either with a shellHook or directly in the `.envrc`, but this is a more project based solution and not every project will need. Working with Tauri will definitely need it.
